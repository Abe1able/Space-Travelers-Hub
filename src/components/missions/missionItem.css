.Active{
    background-color: blue;
    color: #fff;
  }
  
  .Action{
    background-color: silver;
    color: #fff;
  }