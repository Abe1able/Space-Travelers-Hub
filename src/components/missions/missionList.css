.tableSection {
  margin-top: 7rem;
  padding-inline: 2%;
}

.border,
td {
  border: 1px solid #ddd;
  padding: 8px;
}

table {
  border-collapse: collapse;
  width: 100%;
}

th {
  font-weight: 800;
  text-align: left;
  font-size: 20px;
}

.check {
  text-align: center;
}
