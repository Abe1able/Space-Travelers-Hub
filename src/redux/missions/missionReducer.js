import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const GET_MISSION = 'redux/missions/mession';
const apiLink = 'https://api.spacexdata.com/v3/missions';

const renderMission = (res) => res.map((arr) => {
  const { mission_id: missionId, mission_name: missionName, description } = arr;
  return {
    missionId, missionName, description,
  };
});

export const fetchMissions = createAsyncThunk(GET_MISSION, async () => {
  const response = await fetch(apiLink);
  const data = await response.json();
  return renderMission(data);
});

const missionSlice = createSlice({
  name: 'mission',
  initialState: [],
  extraReducers: (builder) => {
    builder.addCase(fetchMissions.fulfilled, (state, action) => {
      let items = state;
      items = action.payload;
      return items;
    });
  },
});

export default missionSlice.reducer;
