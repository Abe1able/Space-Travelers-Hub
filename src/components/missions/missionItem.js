import React from 'react';
import Badge from 'react-bootstrap/Badge';
import Button from 'react-bootstrap/Button';
import { useDispatch } from 'react-redux';
import { joinMission, leavingMission } from '../../redux/missions/missionReducer';

import './missionItem.css';

function MissionItems({ items }) {
  const {
    missionId, missionName, description, reserved,
  } = items;
  const dispatch = useDispatch();

  const joinAmission = () => {
    dispatch(joinMission(missionId));
  };

  const leaveAMission = () => {
    dispatch(leavingMission(missionId));
  };
  return (
    <tr key={missionId}>
      <td className="title">{missionName}</td>
      <td>{description}</td>
      <td className="check">
        <Badge className={reserved ? 'Active' : 'Action'}>
          {reserved ? 'Active Member' : 'Not a Member'}
        </Badge>
      </td>
      <td className="check" width="15%">
        <Button
          id={missionId}
          onClick={reserved ? leaveAMission : joinAmission}
          variant={reserved ? 'outline-danger' : 'outline-secondary'}
        >
          {reserved ? 'Leave Misson' : 'Join Mission'}
        </Button>
      </td>
    </tr>
  );
}

export default MissionItems;
