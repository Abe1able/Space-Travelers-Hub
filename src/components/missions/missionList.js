import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchMissions } from '../../redux/missions/missionReducer';
import MissionItems from './missionItem';

function MissionList() {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.mission);
  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);
  return (
    <section className="tableSection">
      <table>
        <tr>
          <td>Mission</td>
          <td>Description</td>
          <td>Status</td>
          <td>{' '}</td>
        </tr>
        <tbody>
        {
        missions.map((elements) => (

          <MissionItems key={elements.missionId} items={elements} />
        ))
     }
     </tbody>

      </table>
    </section>
  );
}

export default MissionList;
