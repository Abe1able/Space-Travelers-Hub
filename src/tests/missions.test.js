import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import MissionList from '../components/missions/missionList';
import store from '../redux/configureStore';
import { fetchMission, joinMission, leavingMission } from '../redux/missions/missionReducer';

describe('display the mission in the page', () => {
  it('should display all the missions', () => {
    const { container } = render(
      <Provider store={store}>
        <MissionList />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
