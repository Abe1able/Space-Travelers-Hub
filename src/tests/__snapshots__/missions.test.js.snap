// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display the mission in the page should display all the missions 1`] = `
<div>
  <section
    class="tableSection"
  >
    <table>
      <tr>
        <th
          class="border"
          width="10%"
        >
          Mission
        </th>
        <th
          class="border"
          width="60%"
        >
          Description
        </th>
        <th
          class="border"
        >
          Status
        </th>
        <th
          class="border"
        >
           
        </th>
      </tr>
      <tbody />
    </table>
  </section>
</div>
`;
